expo/expo
# [expo-auth-session] Add projectNameForProxy option

# Why

Edit: Keeping the old summary for posterity.

This adds an argument to the auth proxy code so that it can continue to be used with EAS update (new) manifests, which don't always contain the required field since that field has been discontinued.

Closes ENG-1902.

Closes https://github.com/expo/expo/issues/13714.

# How

- ~~Deprecate anything in the library that uses AuthSession proxy (add `@deprecated` to docblocks)~~
- Add more info to the error message when using new manifests and the ID cannot be derived.
- Add new argument.
- ~~Update docs to indicate this deprecation.~~
- Update changelog.

# Test Plan

This is fairly hard to test automatically because all our unit/integration tests are hardcoded for classic manifests: https://github.com/expo/expo/blob/main/packages/jest-expo/src/preset/setup.js#L114 Task: https://linear.app/expo/issue/ENG-4870/un-hardcode-tests-to-be-classic-manifests

Manual test:

1. `export EXPO_BETA=1`
2. `expo init auth-session-test`
3. `expo install expo-auth-session expo-random`
4. Manually point the `expo-auth-session` dependency in package.json to the local version of this package.
5. Add the following code to App.ts:
    ```typescript
    const authUrl = `http://google.com`;
    
    const signIn = async () => {
	    const authRes = await AuthSession.startAsync({
		    authUrl,
	    });
    };
    
    export default function App() {
      return (
        <View style={styles.container}>
          <Text>Open up App.tsx to start working on your app!</Text>
          <Button onPress={signIn} title="DO IT" />
          <StatusBar style="auto" />
        </View>
      );
    }
    ```
6. Run on iOS simulator with `expo start --force-manifest-type=expo-updates`, see error: `Error: Cannot use AuthSession proxy because the project ID is not defined. The AuthSession proxy has been deprecated for modern applications. Prefer AuthRequest (with the useProxy option set to false) in combination with an Expo Development Client build of your application. To continue using the AuthSession proxy during the transition, specify the project ID using the proxyProjectIdOverride option.`
7. Add `projectNameForProxy: '@blah/wat',` option, no longer see error.

# Checklist




# Why (old, for first commit only)

Keeping this for posterity.

The AuthSession proxy requires an "ID" of the form "@account/project" (what we call full name). Historically this was obtained from the `id` field of the classic manifest.

Then, we added project transfer which meant that the `id` field was no longer stable. Instead, we added another field, `originalFullName` which is immutable and the original value of the `id` field.

`originalFullName` is not a field we can fill reliably in development for normal projects using classic manifests since it is only known by the server. We fill it for bare workflow to whatever we fill `id` to. That being said, since we still look at the `id` field as a fallback for development (and old apps) this works.

New manifests don't have an `originalFullName` field or an `id` field (of that format) in development. When served by EAS, they have an `originalFullName` field but it is only there for legacy purposes such as this and doesn't work in development.

So, the AuthSession proxy is broken for EAS/new-manifest apps in development (Expo Go), which is what the proxy was created for (since standalone apps can just use the normal auth flow and redirect).

There are a couple ways we could fix this:
- Create a way to use a different identifier from the manifest, namely the project ID (UUID). The migration path to using project ID is very complex due to the nature of OAuth redirect URIs. Not all services allow a whitelist and instead only allow a single redirect URI, which means that a migration path for that provider would either be a) create a new app on the provider and change the code to use that new app, or b) we allow project-id-based flows to redirect to legacy paths. For both (a) and (b), they defeat the purpose of this library. This was attempted in the PR stack: https://github.com/expo/universe/pull/8170
- Deprecate the proxy and recommend the normal non-proxied auth flows mentioned in the guide (https://docs.expo.dev/guides/authentication). This will require using a dev client for development, but that is okay since there are more and more libraries that require the dev client. This also requires creating a new app on the OAuth provider (since the redirect URI is different) and migrating to that new app.

This PR goes with the second option, but provides a way for users to continue to use the proxy with EAS/new-manifests if they absolutely need to by adding a new argument/option that allows a hardcoded fullName for use by the proxy. Note that if this is used, project transfer will most likely not work as intended.
